README
GOOGLE CHECKOUT MODULE FOR OSCOMMERCE v1.0.4 - 11/30/2006 

INTRODUCTION
============
The Google Checkout module for osCommerce adds Google Checkout as a Checkout Module within osCommerce.
This will allow merchants using osCommerce to offer Google Checkout as an alternate checkout method.
The plugin provides Level 2 integration of Google Checkout with osCommerce.

Plugin features include:
1. Posting shopping carts to Google Checkout
2. Shipping support, now you can use Merchant Calculated Shipping Rates!
3. Tax support
4. User and order updates within osCommerce
5. Order processing using osCommerce Admin UI 


REQUIREMENTS
============
osCommerce v2.2
PHP3 or PHP4 or PHP5 with cURL(libcurl) installed and enabled


INSTALLATION NOTES
==================
* Follow instructions contained in the INSTALLATION file.
* register_globals can be either On or Off. 
Verify the installation from the Admin site and selecting MODULES->PAYMENTS and checking if Google Checkout is listed as a payment option.


SETUP ON ADMIN UI
=================
Select and install the Google Checkout payment module. The following are some of the fields you can update:

1. Enable/Disable: Enable this to use Google Checkout for your site.
2. Merchant ID and Merchant Key:(Mandatory) If any of these are not set and the module is enabled, a disabled (gray) Checkout button appears on the Checkout page. 
   Set these values from your seller Google account under the Settings->Integration tab.
3. Operation Mode: Test your site with Google Checkout's sandbox server before migrating to production. You will need it signup for a separate Google Checkout sandbox account at http://sandbox.google.com/sell. Your sandbox account will have a different Merchant ID and Merchant Key. When you are ready to run against the production server, remember to update your merchant ID and key when migrating.
4. Merchant Calculation Mode of Operation: Sets Merchant calculation URL for Sandbox environment. Could be HTTP or HTTPS. (Checkout production environemnt always requires HTTPS.)
5. Default Values for Real Time Shipping Rates: Set your default values for all merchant calculated shipping rates. This values will be used if for any reason Google Checkout cannot reach your API callback to calculate the shipping price.

Your Google Checkout setup page is correct if, upon viewing it, a non-disabled Google Checkout button appears.


MERCHANT CALCULATED SHIPPING
============================
In order to use this module you must have some Real Time Shipping provider, such as USPS. This Module must be activated and configured in Modules->Shipping. For each enabled module you'll have to set the default values in the Google Checkout Admin UI.
This Value will be used if for any reason Google Checkout cannot reach your API callback to calculate the shipping price. Tha available shipping methods for each shipping provider must be configured in catalog/includes/modules/payment/googlecheckout.php in the mc_shipping_methods parameter. If you want to disable one or more methods, just comment them.
Be aware that if you mix flat rate and real time rates, both will be taken as merchant-calculated-shipping. 


TRACKING USERS AND ORDERS
=========================
In order to provide this support (as required for Level 2 integration), update the API callback request field in the seller account to https://<url-site-url>/catalog/googlecheckout/responsehandler.php . Note that the production Checkout server requires an SSL callback site.
The Sandbox server accepts non SSL callback URLs, http://<url-site-url>/catalog/googlecheckout/responsehandler.php

View your Google Checkout customers and their orders in the Reports tab. 
For each order, the default starting state is PENDING. 
The state changes to PROCESSING and DELIVERED as follows; corresponding messages will be sent out to the Checkout server for order processing.

| Original State | New State  | Action                         |  Customer Notification    |
| PENDING        | PROCESSING | charging the order             |  Processing order message |
| PROCESSING     | DELIVERED  | marking the order for delivery |  Shipped order message    |

Any comments added during state change will be sent to the buyer account page if you have selected the Notify Customer option.

All request and response messages will be logged to the file catalog/googlecheckout/response_message.log.
In case an order payment is declined or an order is cancelled(by Google or otherwise), you can add code to catalog/googlecheckout/responsehandler.php for any specific actions you need to take.


PROJECT HOME
============
To check out the latest release or to report issues, go to http://code.google.com/p/google-checkout-oscommerce


GROUP DISCUSSIONS
=================
To meet other developers and merchants who have integrated Google Checkout with osCommerce and to discuss any topics related to integration of Google Checkout with osCommerce, go to http://forums.oscommerce.com/index.php?showtopic=229637

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
MOST COMMON MISTAKES
====================
1. Make sure you set the file attribute to 777 for /googlecheckout/response_error.log and /googlecheckout/response_message.log files.
2. Set your Google callback url to https://<url-site-url>/googlecheckout/responsehandler.php
  In Sandbox, HTTPS is not required.
  In Production mode, HTTPS is required.
  Set the correct option in the Google Checkout Admin UI
3. Make sure you are using the correct combination of Merchant ID and Merchant Key. Remember that Sandbox and Production Mode have different ones.
  

TROUBLE SHOOTING
================
1. Problem: /public_html/googlecheckout/response_message.log) [function.fopen]: failed to open stream: Permission denied.
   Solution: Set the file attribute to 777 for /googlecheckout/response_error.log and /googlecheckout/response_message.log files.
2. Problem: Test order shows up in Google but not admin.
   Solution: There is an error somewhere in the file /googlecheckout/responsehandler.php or you have set a worng API callback function in your seller Google account under the Settings->Integration tab.
3. Problem: <error-message>Malformed URL component: expected id: (\d{10})|(\d{15}), but got 8***********4 </error-message>
   Solution: You have an extra space after your Google merchant id. Go to Admin->payment.  Edit Googlecheckout module.  Extra space will disappear. Click update button.
4. Problem: <error-message>No seller found with id 7************8</error-message>
   Solution: Wrong merchant id.  Sandbox merchant id can only be use with sandbox accounts.  Sandbox and Live mode use different merchant id. 
5. Problem: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderE xception: unable to find valid certification path to requested target
   Solution: Your SSL certificate is not accepted by Google Checkout.
6. Problem: <error-message>Bad Signature on Cart</error-message>
   Solution: Incorrect Merchant key.
7. Problem: (/public_html/googlecheckout/response_error.log) Tue Nov 28 8:56:21 PST 2006:- Shopping cart not obtained from session. 
   Solution: Set to False admin->configuration->session->Prevent Spider Sessions configuration (Thx dawnmariegifts, beta tester)
   Side effects: You'll see spiders as active users. Working on a permanent fix.
   
    
KNOWN BUGS - (Report bugs at http://forums.oscommerce.com/index.php?showtopic=229637)
==========


CHANGE LOG
==========
09/26/2006 v1.0.0 (Google Checkout team)
           - Initial release. 

09/27/2006 v1.0.1 (Google Checkout team)
           - Updated the original module to run on either PHP4 or PHP5.

10/05/2006 v1.0.1b (Google Checkout team)
           - Step-by-step installation instructions included.

10/18/2006 v1.0.2 (Google Checkout team)
           - Fixed minor bugs in responsehandler.php and orders.php files
  
11/17/2006 v1.0.3 (ropu)
           - Add support for Merchant Calculated Shipping Rates.
           - Fixed minor bugs in responsehandler.php and orders.php files
           - Change the XML parser and builder
           - Removed getallheader() function
           - Fixed wrong Qty in Admin UI
           - Fixed modules not saving their settings
           - Fixed Notify Customer option

11/30/2006 v1.0.4 (ropu)
           - Add state change, Risk and charge notification into the Admin UI
           - Fix Shopping cart not obtained from session. See TROUBLE SHOOTING.


CHANGELOG
CHANGELOG
==========
09/26/2006 v1.0.0 (Google Checkout team)
   - Initial release. 

09/27/2006 v1.0.1 (Google Checkout team)
   - Updated the original module to run on either PHP4 or PHP5.

10/05/2006 v1.0.1b (Google Checkout team)
   - Step-by-step installation instructions included.

10/18/2006 v1.0.2 (Google Checkout team)
   - Fixed minor bugs in responsehandler.php and orders.php files
  
11/17/2006 v1.0.3 (ropu)
   - Add support for Merchant Calculated Shipping Rates.
   - Fixed minor bugs in responsehandler.php and orders.php files
   - Change the XML parser and builder
   - Removed getallheader() function
   - Fixed wrong Qty in Admin UI
   - Fixed modules not saving their settings
   - Fixed Notify Customer option

12/04/2006 v1.0.4 (ropu)
   - Add order-state-change, risk-information and charge-amount notification into the Admin UI
   - Fix Shopping cart not obtained from session. See TROUBLE SHOOTING.

01/10/2007 v1.1.0b3 (rszrama)
   - Bugfix compilation so people stop downloading the old code till 1.1.0 comes out!
   - Read v1-1-0b2.txt for more information.
   
01/12/2007 v1.1.0b4 (ropu)
   - Fix <tax-table-selector> strict validation.
   
02/26/2007 v1.2 (ropu)
   - Add multisocket feature for merchant-calculations (alfa)(optional)
   - Different algorithm to retrieve quotes
   - Add Google Analytics Support
   - Add support for PHP CGI installations
   - Add in UPSXML methods by default
   - Add user and password for Google Checkout buyers
   - Items retrieved from Merchant-private-item-data instead of session.
   - Many bug fixes

03/05/2007 v1.2 RC3 (ropu)
   - Fix gray button when Tax Class selected bug (Thx BlessIsaacola)
   
04/10/2007 v1.3RC1 (ropu)
   - Add tracking data to the Admin UI Orders
   - Fixed SSL issue with Google Analytics feature
   - International Shipping Features
   - Restricting Shipping Options for Post Office (P.O.) Box Addresses feature
   - International Tax Features
   - Selecting a Rounding Policy for Tax Calculations
   - Fixed Tax for zones
   - Fixed Tax for products
   - Added support for All Areas Zones
   - Add a configuration to disable Google Checkout Button when are virtual good in the cart
   		(double check http://checkout.google.com/seller/policies.html#4)
   - Disable multisocket Option :(
   
04/18/2007 v1.3RC2 (ropu)
   - Support for UK merchants 
     http://code.google.com/apis/checkout/developer/release_notes.html (apr13)
   - Docs update
   - Minor bug fix
   - Added Shipping Generator Tool in the stardard package (Optional)
   - Flat rate shipping bug fixed
   - Improved tax and shipping restrictions
   
06/20/2007 v1.4 apha (ropu)
   - Refactor all code using PHP Sample Code as the base library (under googlecheckout/library)
     - Handles Logging, added log levels (error, request, response)
       - Configuration from UI (comming soon), needed??
     - Handles Basic authentication, add support for IIS (See TROUBLESHOOTING)
     - Handles all XML parsing and building (New xml parsed used)
   - Added OT support (low order fees, group discounts and any other custom OT module)
     - feedback needed. Calculating tax for this, still an issue.
   - Add warning messages when GC button is disabled.
   - Add stack message in Admin UI with GC orders state changes and errors
   - Added configuration to also use OSC email system for GC orders
   - Added shipping generator (googlecheckout/shiping_generator/) (optional)
   - Added a metric tester for shipping times feedback (optional)
   - Added 3rd party tracking
   - Added a coustom Continue shopping url with GC bought items.
   - Moved logs to a .htacces protected dir (googlecheckout/logs)
   - Added Correct HTTP status (2xx, 4xx, 5xx) return when errors occurrs with the correct message
   - All strings are defined in the language include (only english right now)
   - Use DEFAULT_CURRENCY as the currency in GC cart, not buyers session one
   - Added version number to all files
   - Added GC Module Installed and Configured Version check.
   - Slipt README and CHANGELOG
 	 
17/07/2007 v1.4beta1 (ropu)
    - Refactor new-order-notification handling
    - Merged the GC return page with the OSC default one, checkout_success.php 
       used as continue-shopping URL. Buyer will return with his logged in to OSC
    - Digital Delivery support
    - Support Google Refunds and Cancellations
    - Fixed GC Invoicing issue, items will be added to the order details
    - If the customers e-mail address already exists in OSC then add the
        order placed to their account instead of creating a new customer
    - If the GC exits in OSC but uses a different email then add the
        order placed to their account instead of creating a new customer.
    - Check if new customers have <email-allowed> and activate newsletters
 	  - OT tax support
    - Existing OSC customer accounts are updated when GC orders processed
    - No more need to know admin folders name
    - Add GC Module Version Control to avoid mix of configured and installed modules errors
    - Moved Shipping quotes definition from includes/modules/payment/googlecheckout.php
        to googlecheckout/shipping_methods.php
    - Added OSC Order# to the GC Order when state changes from Pending to Processing
    - Added index.html to googlecheckout/* dirs.
    - Upgraded to use OSCommerce v2.2RC1

09/27/2007 v1.4.5_CCS (ropu)
    - Added Carrier Calculated Shipping Support (CCS)!!
      (http://code.google.com/apis/checkout/developer/Google_Checkout_XML_API_Carrier_Calculated_Shipping.html)
    - Added Sandbox and Production id/key config in Admin UI
    - Added SANDBOX string to the payment method in order page for orders processed
       with GC using sandbox.
    - Added a warning under the GC Button when in Sandbox mode.
    - Added a warning in the Payment Modules page in the Admin UI when in Sandbox mode.    
    - Disalbe merchant Calculation Shipping methods from the Admin UI by setting 
       Default value to 0.

catalog/admin/modules.php
<?php
/*
$Id: modules.php 3 2006-05-27 04:59:07Z user $

  osCMax Power E-Commerce
  http://oscdox.com

  Copyright 2006 osCMax

  Released under the GNU General Public License
*/

  require('includes/application_top.php');
  // ** GOOGLE CHECKOUT **
  function gc_makeSqlString($str) {
    $single_quote = "'";
    $escaped_str = addcslashes(stripcslashes($str), "'\"\\\0..\37!@\177..\377");
    return ($single_quote.$escaped_str.$single_quote);
  }
  // **END GOOGLE CHECKOUT**

  $set = (isset($HTTP_GET_VARS['set']) ? $HTTP_GET_VARS['set'] : '');

  if (tep_not_null($set)) {
    switch ($set) {
      case 'shipping':
        $module_type = 'shipping';
        $module_directory = DIR_FS_CATALOG_MODULES . 'shipping/';
        $module_key = 'MODULE_SHIPPING_INSTALLED';
        define('HEADING_TITLE', HEADING_TITLE_MODULES_SHIPPING);
        break;
      case 'ordertotal':
        $module_type = 'order_total';
        $module_directory = DIR_FS_CATALOG_MODULES . 'order_total/';
        $module_key = 'MODULE_ORDER_TOTAL_INSTALLED';
        define('HEADING_TITLE', HEADING_TITLE_MODULES_ORDER_TOTAL);
        break;
      case 'payment':
      default:
        $module_type = 'payment';
        $module_directory = DIR_FS_CATALOG_MODULES . 'payment/';
        $module_key = 'MODULE_PAYMENT_INSTALLED';
        define('HEADING_TITLE', HEADING_TITLE_MODULES_PAYMENT);
        break;
    }
  }

  $action = (isset($HTTP_GET_VARS['action']) ? $HTTP_GET_VARS['action'] : '');

  if (tep_not_null($action)) {
    switch ($action) {
      case 'save':
      // ** GOOGLE CHECKOUT **      
        // fix configuration no saving -
      	reset($HTTP_POST_VARS['configuration']);
        // end fix
      // ** END GOOGLE CHECKOUT **      
        while (list($key, $value) = each($HTTP_POST_VARS['configuration'])) {
// BOF: LINES ADDED
        // ** GOOGLE CHECKOUT **    
          // Checks if module is of type google checkout and also verfies if this configuration is 
          // for the check boxes for the shipping options   				
          if( is_array( $value ) ){
          $value = implode( ", ", $value);
		  $value = ereg_replace (", --none--", "", $value);
              }
        // ** END GOOGLE CHECKOUT **
// EOF: LINES ADDED
          tep_db_query("update " . TABLE_CONFIGURATION . " set configuration_value = '" . $value . "' where configuration_key = '" . $key . "'");
        }
        tep_redirect(tep_href_link(FILENAME_MODULES, 'set=' . $set . '&module=' . $HTTP_GET_VARS['module']));
        break;
      case 'install':
      case 'remove':
        $file_extension = substr($PHP_SELF, strrpos($PHP_SELF, '.'));
        $class = basename($HTTP_GET_VARS['module']);
        if (file_exists($module_directory . $class . $file_extension)) {
          include($module_directory . $class . $file_extension);
          $module = new $class;
          if ($action == 'install') {
            $module->install();
          } elseif ($action == 'remove') {
            $module->remove();
          }
        }
        tep_redirect(tep_href_link(FILENAME_MODULES, 'set=' . $set . '&module=' . $class));
        break;
    }
  }
?>
<!doctype html public "-//W3C//DTD HTML 4.01 Transitional//EN">
<html <?php echo HTML_PARAMS; ?>>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=<?php echo CHARSET; ?>">
<title><?php echo TITLE; ?></title>
<link rel="stylesheet" type="text/css" href="includes/stylesheet.css">
<script language="javascript" src="includes/general.js"></script>
</head>
<body marginwidth="0" marginheight="0" topmargin="0" bottommargin="0" leftmargin="0" rightmargin="0" bgcolor="#FFFFFF">
<!-- header //-->
<?php require(DIR_WS_INCLUDES . 'header.php'); ?>
<!-- header_eof //-->

<!-- body //-->
<table border="0" width="100%" cellspacing="2" cellpadding="2">
  <tr>
    <td width="<?php echo BOX_WIDTH; ?>" valign="top"><table border="0" width="<?php echo BOX_WIDTH; ?>" cellspacing="1" cellpadding="1" class="columnLeft">
<!-- left_navigation //-->
<?php require(DIR_WS_INCLUDES . 'column_left.php'); ?>
<!-- left_navigation_eof //-->
    </table></td>
<!-- body_text //-->
    <td width="100%" valign="top"><table border="0" width="100%" cellspacing="0" cellpadding="2">
      <tr>
        <td width="100%"><table border="0" width="100%" cellspacing="0" cellpadding="0">
          <tr>
            <td class="pageHeading"><?php echo HEADING_TITLE; ?></td>
            <td class="pageHeading" align="right"><?php echo tep_draw_separator('pixel_trans.gif', HEADING_IMAGE_WIDTH, HEADING_IMAGE_HEIGHT); ?></td>
          </tr>
        </table></td>
      </tr>
      <tr>
        <td><table border="0" width="100%" cellspacing="0" cellpadding="0">
          <tr>
            <td valign="top"><table border="0" width="100%" cellspacing="0" cellpadding="2">
              <tr class="dataTableHeadingRow">
                <td class="dataTableHeadingContent"><?php echo TABLE_HEADING_MODULES; ?></td>
                <td class="dataTableHeadingContent" align="right"><?php echo TABLE_HEADING_SORT_ORDER; ?></td>
                <td class="dataTableHeadingContent" align="right"><?php echo TABLE_HEADING_ACTION; ?>&nbsp;</td>
              </tr>
<?php
  $file_extension = substr($PHP_SELF, strrpos($PHP_SELF, '.'));
  $directory_array = array();
  if ($dir = @dir($module_directory)) {
    while ($file = $dir->read()) {
      if (!is_dir($module_directory . $file)) {
        if (substr($file, strrpos($file, '.')) == $file_extension) {
          $directory_array[] = $file;
        }
      }
    }
    sort($directory_array);
    $dir->close();
  }

  $installed_modules = array();
  for ($i=0, $n=sizeof($directory_array); $i<$n; $i++) {
    $file = $directory_array[$i];

    include(DIR_FS_CATALOG_LANGUAGES . $language . '/modules/' . $module_type . '/' . $file);
    include($module_directory . $file);

    $class = substr($file, 0, strrpos($file, '.'));
    if (tep_class_exists($class)) {
      $module = new $class;
      if ($module->check() > 0) {
        if ($module->sort_order > 0) {
          $installed_modules[$module->sort_order] = $file;
        } else {
          $installed_modules[] = $file;
        }
      }

      if ((!isset($HTTP_GET_VARS['module']) || (isset($HTTP_GET_VARS['module']) && ($HTTP_GET_VARS['module'] == $class))) && !isset($mInfo)) {
        $module_info = array('code' => $module->code,
                             'title' => $module->title,
                             'description' => $module->description,
                             'status' => $module->check());

        $module_keys = $module->keys();

        $keys_extra = array();
        for ($j=0, $k=sizeof($module_keys); $j<$k; $j++) {
          $key_value_query = tep_db_query("select configuration_title, configuration_value, configuration_description, use_function, set_function from " . TABLE_CONFIGURATION . " where configuration_key = '" . $module_keys[$j] . "'");
          $key_value = tep_db_fetch_array($key_value_query);

          $keys_extra[$module_keys[$j]]['title'] = $key_value['configuration_title'];
          $keys_extra[$module_keys[$j]]['value'] = $key_value['configuration_value'];
          $keys_extra[$module_keys[$j]]['description'] = $key_value['configuration_description'];
          $keys_extra[$module_keys[$j]]['use_function'] = $key_value['use_function'];
          $keys_extra[$module_keys[$j]]['set_function'] = $key_value['set_function'];
        }

        $module_info['keys'] = $keys_extra;

        $mInfo = new objectInfo($module_info);
      }

      if (isset($mInfo) && is_object($mInfo) && ($class == $mInfo->code) ) {
        if ($module->check() > 0) {
          echo '              <tr id="defaultSelected" class="dataTableRowSelected" onmouseover="rowOverEffect(this)" onmouseout="rowOutEffect(this)" onclick="document.location.href=\'' . tep_href_link(FILENAME_MODULES, 'set=' . $set . '&module=' . $class . '&action=edit') . '\'">' . "\n";
        } else {
          echo '              <tr id="defaultSelected" class="dataTableRowSelected" onmouseover="rowOverEffect(this)" onmouseout="rowOutEffect(this)">' . "\n";
        }
      } else {
        echo '              <tr class="dataTableRow" onmouseover="rowOverEffect(this)" onmouseout="rowOutEffect(this)" onclick="document.location.href=\'' . tep_href_link(FILENAME_MODULES, 'set=' . $set . '&module=' . $class) . '\'">' . "\n";
      }
?>
                <td class="dataTableContent"><?php echo $module->title; ?></td>
                <td class="dataTableContent" align="right"><?php if (is_numeric($module->sort_order)) echo $module->sort_order; ?></td>
                <td class="dataTableContent" align="right"><?php if (isset($mInfo) && is_object($mInfo) && ($class == $mInfo->code) ) { echo tep_image(DIR_WS_IMAGES . 'icon_arrow_right.gif'); } else { echo '<a href="' . tep_href_link(FILENAME_MODULES, 'set=' . $set . '&module=' . $class) . '">' . tep_image(DIR_WS_IMAGES . 'icon_info.gif', IMAGE_ICON_INFO) . '</a>'; } ?>&nbsp;</td>
              </tr>
<?php
    }
  }

  ksort($installed_modules);
  $check_query = tep_db_query("select configuration_value from " . TABLE_CONFIGURATION . " where configuration_key = '" . $module_key . "'");
  if (tep_db_num_rows($check_query)) {
    $check = tep_db_fetch_array($check_query);
    if ($check['configuration_value'] != implode(';', $installed_modules)) {
      tep_db_query("update " . TABLE_CONFIGURATION . " set configuration_value = '" . implode(';', $installed_modules) . "', last_modified = now() where configuration_key = '" . $module_key . "'");
    }
  } else {
    tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, date_added) values ('Installed Modules', '" . $module_key . "', '" . implode(';', $installed_modules) . "', 'This is automatically updated. No need to edit.', '6', '0', now())");
  }
?>
              <tr>
                <td colspan="3" class="smallText"><?php echo TEXT_MODULE_DIRECTORY . ' ' . $module_directory; ?></td>
              </tr>
            </table></td>
<?php
  $heading = array();
  $contents = array();

  switch ($action) {
    case 'edit':
      $keys = '';
      reset($mInfo->keys);
      while (list($key, $value) = each($mInfo->keys)) {
        $keys .= '<b>' . $value['title'] . '</b><br>' . $value['description'] . '<br>';

        if ($value['set_function']) {
          eval('$keys .= ' . $value['set_function'] . "'" . $value['value'] . "', '" . $key . "');");
        } else {
          $keys .= tep_draw_input_field('configuration[' . $key . ']', $value['value']);
        }
        $keys .= '<br><br>';
      }
      $keys = substr($keys, 0, strrpos($keys, '<br><br>'));

      $heading[] = array('text' => '<b>' . $mInfo->title . '</b>');

      $contents = array('form' => tep_draw_form('modules', FILENAME_MODULES, 'set=' . $set . '&module=' . $HTTP_GET_VARS['module'] . '&action=save'));
      $contents[] = array('text' => $keys);
      $contents[] = array('align' => 'center', 'text' => '<br>' . tep_image_submit('button_update.gif', IMAGE_UPDATE) . ' <a href="' . tep_href_link(FILENAME_MODULES, 'set=' . $set . '&module=' . $HTTP_GET_VARS['module']) . '">' . tep_image_button('button_cancel.gif', IMAGE_CANCEL) . '</a>');
      break;
    default:
      $heading[] = array('text' => '<b>' . $mInfo->title . '</b>');

      if ($mInfo->status == '1') {
        $keys = '';
        reset($mInfo->keys);
        while (list(, $value) = each($mInfo->keys)) {
          $keys .= '<b>' . $value['title'] . '</b><br>';
          if ($value['use_function']) {
            $use_function = $value['use_function'];
            if (ereg('->', $use_function)) {
              $class_method = explode('->', $use_function);
              if (!is_object(${$class_method[0]})) {
                include(DIR_WS_CLASSES . $class_method[0] . '.php');
                ${$class_method[0]} = new $class_method[0]();
              }
              $keys .= tep_call_function($class_method[1], $value['value'], ${$class_method[0]});
            } else {
              $keys .= tep_call_function($use_function, $value['value']);
            }
          } else {
            $keys .= $value['value'];
          }
          $keys .= '<br><br>';
        }
        $keys = substr($keys, 0, strrpos($keys, '<br><br>'));

        $contents[] = array('align' => 'center', 'text' => '<a href="' . tep_href_link(FILENAME_MODULES, 'set=' . $set . '&module=' . $mInfo->code . '&action=remove') . '">' . tep_image_button('button_module_remove.gif', IMAGE_MODULE_REMOVE) . '</a> <a href="' . tep_href_link(FILENAME_MODULES, 'set=' . $set . (isset($HTTP_GET_VARS['module']) ? '&module=' . $HTTP_GET_VARS['module'] : '') . '&action=edit') . '">' . tep_image_button('button_edit.gif', IMAGE_EDIT) . '</a>');
        $contents[] = array('text' => '<br>' . $mInfo->description);
        $contents[] = array('text' => '<br>' . $keys);
      } else {
        $contents[] = array('align' => 'center', 'text' => '<a href="' . tep_href_link(FILENAME_MODULES, 'set=' . $set . '&module=' . $mInfo->code . '&action=install') . '">' . tep_image_button('button_module_install.gif', IMAGE_MODULE_INSTALL) . '</a>');
        $contents[] = array('text' => '<br>' . $mInfo->description);
      }
      break;
  }

  if ( (tep_not_null($heading)) && (tep_not_null($contents)) ) {
    echo '            <td width="25%" valign="top">' . "\n";

    $box = new box;
    echo $box->infoBox($heading, $contents);

    echo '            </td>' . "\n";
  }
?>
          </tr>
        </table></td>
      </tr>
    </table></td>
<!-- body_text_eof //-->
  </tr>
</table>
<!-- body_eof //-->

<!-- footer //-->
<?php require(DIR_WS_INCLUDES . 'footer.php'); ?>
<!-- footer_eof //-->
<br>
</body>
</html>
<?php require(DIR_WS_INCLUDES . 'application_bottom.php'); ?>
